### PR TITLE
PP-8947 emit FeeIncurredEvent

### PR DESCRIPTION
**Pull request by SandorArpa**
## WHAT YOU DID
- emit FeeIncurredEvent after successful capture when charge has more than 1 FeeEntity
- do not send fee details with CaptureConfirmed if charge has more than 1 FeeEntity

with @nimalank7

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java [hunk 1]
```diff
@@ -173,4 +192,16 @@ private ChargeStatus determineNextStatus(CaptureResponse operationResponse) {
             return CAPTURED;
         }
     }
+
+    private void sendToEventQueue(Event event) {
+        try {
+            eventService.emitEvent(event, false);
+            logger.info("Payout event sent to event queue",
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java]: Fee incurred event ?

File: src/test/java/uk/gov/pay/connector/events/model/charge/CaptureConfirmedTest.java [hunk 2]
```diff
@@ -45,6 +46,31 @@ public void serializesEventDetailsGivenChargeEvent() throws JsonProcessingExcept
         assertThat(actual, hasJsonPath("$.event_details.captured_date", equalTo("2018-03-12T16:25:02.123456Z")));
     }
 
+    @Test
+    public void serializesEventDetailsGivenChargeEventWithMultipleFees() throws JsonProcessingException {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/events/model/charge/CaptureConfirmedTest.java]: Could we rename this to something like below or more appropriate 
```suggestion
    public void shouldSerializeEventDetailsWithoutFeeGivenChargeWithMultipleFees() throws JsonProcessingException {
```

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardServiceTest.java [hunk 3]
```diff
@@ -57,6 +59,21 @@ protected ChargeEntity createNewChargeWith(Long chargeId, ChargeStatus status) {
 
     }
 
+    protected ChargeEntity createNewChargeWithFees(String provider, Long chargeId, ChargeStatus status, String gatewayTransactionId) {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardServiceTest.java:62]: Is this needed at all ? i.e., are the fees added to charge here used in tests ?

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureServiceTest.java [hunk 4]
```diff
@@ -154,7 +154,14 @@ private void worldpayWillRespondWithSuccess() {
 
     private void stripeWillRespondWithSuccess() {
         when(mockedPaymentProvider.capture(any())).thenReturn(
-                fromBaseCaptureResponse(BaseCaptureResponse.fromTransactionId(randomUUID().toString(), STRIPE), PENDING, 50L, List.of(Fee.of(FeeType.TRANSACTION, 50L))));
+                fromBaseCaptureResponse(BaseCaptureResponse.fromTransactionId(randomUUID().toString(), STRIPE), PENDING, 50L,
+                        List.of(Fee.of(FeeType.TRANSACTION, 50L))));
+    }
+
+    private void stripeWillRespondWithSuccessV2() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureServiceTest.java]: could possibly name it differently to be clear

```suggestion
    private void stripeWillRespondWithSuccessAndAdditionalFees() {
```

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardServiceTest.java (hunk 3 above)
Comment (SandorArpa) [reply src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardServiceTest.java:62]: Yes, we are testing whether the EventService is called in `doCapture_shouldCaptureAChargeForStripeAccountV2`. For that we need multiple fees on the charge entity.

File: src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java [hunk 5]
```diff
@@ -121,9 +132,17 @@ public void processGatewayCaptureResponse(String chargeId, String oldStatus, Cap
         ChargeEntity charge = chargeService.updateChargePostCapture(chargeId, nextStatus);
 
         captureResponse.getFeeList().ifPresent(feeList ->
-            feeList.forEach(fee -> persistFee(charge, fee))
+                feeList.forEach(fee -> persistFee(charge, fee))
         );
 
+        if (charge.getFees().size() > 1) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java]: Sorry missed this one.

Can we move this to block `captureResponse.getFeeList().ifPresent(feeList ->`  as feeList won't be available for all payment providers ?

File: src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java [hunk 6]
```diff
@@ -173,4 +194,16 @@ private ChargeStatus determineNextStatus(CaptureResponse operationResponse) {
             return CAPTURED;
         }
     }
+
+    private void sendToEventQueue(Event event) {
+        try {
+            eventService.emitEvent(event, false);
+            logger.info("Fee incurred event sent to event queue",
+                    kv(LEDGER_EVENT_TYPE, event.getEventType()));
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java]: Can we also add PAYMENT_EXTERNAL_ID ?

File: src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java [hunk 7]
```diff
@@ -173,4 +194,16 @@ private ChargeStatus determineNextStatus(CaptureResponse operationResponse) {
...
+        } catch(QueueException e) {
+            logger.error(format("Error sending fee incurred event to event queue: exception [%s]", e.getMessage()),
+                    kv(LEDGER_EVENT_TYPE, event.getEventType()),
+                    kv(GATEWAY_PAYOUT_ID, event.getResourceExternalId()));
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java]:
```suggestion
                    kv(PAYMENT_EXTERNAL_ID, event.getResourceExternalId()));
```

File: src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java [hunk 8]
```diff
@@ -27,11 +31,15 @@
 import java.util.Optional;
 
 import static java.lang.String.format;
+import static net.logstash.logback.argument.StructuredArguments.kv;
 import static uk.gov.pay.connector.charge.model.domain.ChargeStatus.CAPTURED;
 import static uk.gov.pay.connector.charge.model.domain.ChargeStatus.CAPTURE_APPROVED_RETRY;
 import static uk.gov.pay.connector.charge.model.domain.ChargeStatus.CAPTURE_ERROR;
 import static uk.gov.pay.connector.charge.model.domain.ChargeStatus.CAPTURE_SUBMITTED;
 import static uk.gov.pay.connector.gateway.CaptureResponse.ChargeState.PENDING;
+import static uk.gov.service.payments.logging.LoggingKeys.CONNECT_ACCOUNT_ID;
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java]: Unused import ?

**Review by kbottla — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/pay/connector/events/eventdetails/charge/CaptureConfirmedEventDetails.java [hunk 9]
```diff
@@ -24,6 +24,14 @@ public CaptureConfirmedEventDetails(ZonedDateTime gatewayEventDate, ZonedDateTim
     }
 
     public static CaptureConfirmedEventDetails from(ChargeEventEntity chargeEvent) {
+        if (chargeEvent.getChargeEntity().getFees().size() > 1) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/events/eventdetails/charge/CaptureConfirmedEventDetails.java:27]: I am assuming getFees() will always be available here ? Empty list if payment provider is worldpay or some entries for stripe.

Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/events/eventdetails/charge/CaptureConfirmedEventDetails.java:27]: Yes, same as getEvents()

**Review by kbottla — APPROVED** [review]
LGTM!

**Pull request merged**